cf-custom-resources/lib/custom-domain-app-runner.js

```diff
@@ -234,25 +241,128 @@ async function validateCertForDomain(serviceARN, domainName) {
         });
 
         lastDomainStatus = domain.Status;
-        if (lastDomainStatus !== DOMAIN_STATUS_PENDING_VERIFICATION) {
+
+        if (!domainValidationRecordReady(domain)) {
             await sleep(3000);
             continue;
         }
+
         // Upsert all records needed for certificate validation.
         const records = domain.CertificateValidationRecords;
+        let promises = [];
         for (const record of records) {
-            await updateCNAMERecordAndWait(record.Name, record.Value, appHostedZoneID, "UPSERT").catch(err => {
-                throw new Error(`update validation records for domain ${domainName}: ` + err.message);
-            });
+            promises.push(
+                updateCNAMERecordAndWait(record.Name, record.Value, appHostedZoneID, "UPSERT").catch(err => {
+                    throw new Error(`update validation records for domain ${domainName}: ` + err.message);
+                })
+            );
         }
-        break;
+        return Promise.all(promises);
     }
 
     if (i === ATTEMPTS_WAIT_FOR_PENDING) {
         throw new Error(`update validation records for domain ${domainName}: fail to wait for state ${DOMAIN_STATUS_PENDING_VERIFICATION}, stuck in ${lastDomainStatus}`);
     }
 }
 
+function domainValidationRecordReady(domain) {
+    if (domain.Status === DOMAIN_STATUS_PENDING_VERIFICATION) {
+        return true;
+    }
+
+    if (domain.Status === DOMAIN_STATUS_ACTIVE && domain.CertificateValidationRecords && domain.CertificateValidationRecords.length === 3) {
+        // There are one known scenarios where status could be ACTIVE right after it's associated:
+        // When the domain just got deleted and added again. In this case, even though the validation records could
+        // have been deleted, the previously successful validation results are still cached. Because of the cache,
+        // the domain will show to be ACTIVE immediately after it's associated , although the validation records are not
+        // there anymore.
+        // In this case, the status won't transit to PENDING_VERIFICATION, so we need to check whether the validation
+        // records are ready by counting if there are three of them.
```

Comment:
Maybe instead of doing this, it would be better to have the function docstring.

Reply:
Why do we care about the domain status? Can we just check if the validation record length is 3 or not?

Reply:
I think checking the `PENDING_VERIFICATION` status conforms to the API better. Checking the number of records also works, but I feel like it's suboptimal because it makes assumptions on the API, hence I only did this when I had to.

Reply:
addressed!